modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +162,141 @@ impl VulnerabilityService {
 
         Ok(result.rows_affected)
     }
+
+    pub async fn analyze_purls<C>(
+        &self,
+        purls: Vec<&str>,
+        connection: &C,
+    ) -> Result<HashMap<String, Vec<VulnerabilityDetails>>, Error>
+    where
+        C: ConnectionTrait + StreamTrait,
+    {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let queries: Vec<Result<String, Error>> = purls
+            .iter()
+            .map(|p| {
+                let purl = match Purl::from_str(p) {
+                    Ok(purl) => purl,
+                    Err(e) => return Err(Error::Purl(e)),
+                };
```

Comment:
The `?` is your friend here
```suggestion
                let purl = Purl::from_str(p)?;
```